Added ldc.simd

jkrempus opened the pull request:
This pull request adds a new module named ldc.simd. This module contains some SIMD templates that can't be implemented using x86 builtins. These functions are needed to implement  libraries such as std.simd. 

The functions in this module are:
- shufflevector: This function provides access to [LLVM's shufflevector instruction](http://llvm.org/docs/LangRef.html#i_shufflevector). Usage:

``` d
int4 a, b;
int4 c = shufflevector!(int4, 0, 2, 4, 6)(a, b);
```
- insertelement: This function provides access to [LLVM's insertelement instruction](http://llvm.org/docs/LangRef.html#i_insertelement). Usage:

``` d
int4 a;
int k;
int4 b = insertelement!(int4, 2)(a, k);
```
- extractelement: This function provides access to [LLVM's extractelement instruction](http://llvm.org/docs/LangRef.html#i_extractelement). Usage:

``` d
int4 a;
int k = extractelement!(int4, 2)(a);
```
- loadUnaligned: Loads a vector from an unaligned pointer. Usage:

``` d
int* p;
int4 a = loadUnaligned!int4(p);
```
- equalMask, notEqualMask, greaterMask, greaterOrEqualMask. These functions perform an element-wise comparison between two vectors and return a vector with signed integral elements. The number of elements in the returned vector and their size is the same as in parameter vectors. If the condition in the name of the function holds for elements of the parameter vectors at a given index, all bits of the element of the result at that index are set to 1, otherwise the element of the result is zero.  Usage:

``` d
float4 a, b;
// equivalent to doing c.array[i] = a.array[i] > b.array[i] ? 0xffff_ffff : 0 for all elements
int4 c = greaterMask!int4(a, b);
```

I have a fork of TurkeyMan's std.simd that is implemented using this module and the intrinsics in ldc.gccbuiltins_x86 (and no other language extensions)  [here](https://github.com/jerro/phobos/blob/use-ldc.simd/std/simd.d). I'm testing it with [this file](https://github.com/jerro/std.simd-tests/blob/master/test.d). The tests for all the functions that use ldc.simd pass.

jkrempus committed:
Added ldc.simd

dnadlinger commented:
Can you please add your comments from the pull request description directly to the source in the form of DDoc comments? I will try to turn the intrinsics comments into proper DDoc ones and upload a DDoc build of the ldc.\* modules later.

The code looks fine to me.

jkrempus committed:
Added documentation to ldc.simd

Also, fixed a bug in cmpMask.

jkrempus commented:
I added DDoc comments and fixed a bug in cmpMask (apparently my tests didn't distinguish between greater and greater or equal). Should I also add the file to the build system and update the submodule in the LDC repository, or will you do that?

dnadlinger commented:
As I am somewhat swamped right now, it would be great if you could do it. I think you will need me to merge this first, though, in order to be able reference the commit in the submodule.

In any case, a final nitpick: Can you please break the super long lines in the comments down to ~80 chars. Other than that, looks great.

Oh, and again thanks for all the work you are putting into this!

jkrempus commented:
Done.

And thank _you_ for all the work you have put in LDC :D
